### PR TITLE
[DOCS] Adding create index snapshot API docs

### PR DESCRIPTION
**Pull request by lockewritesdocs**
Adds a new page for the [Create index snapshot API](https://elasticsearch_58519.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/create-snapshot-api.html).

Relates to [#56069](https://github.com/elastic/elasticsearch/issues/56069)

* 7.x backport #58692 
* 7.8 backport #58694

### CODE REVIEW

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:122]: Snippet tests help ensure our docs are correct. Generally, we don't use `// TEST[skip:...` unless absolutely necessary.  In this case,  I think the skipped test is hiding an incorrect response.

I suggested a valid response below. Here's how you can get it yourself next time:

* Use a command like `./gradlew :docs:integTestRunner --tests "*create-snapshot-api*"` to test the snippets in the file.
* Use `// TESTRESPONSE[s/.../.../]` to selectively replace some properties in the response.

This is covered in more depth in the docs README: https://github.com/elastic/elasticsearch/tree/master/docs

```suggestion
[source,console-result]
----
{
  "snapshot": {
    "snapshot": "snapshot_2",
    "uuid": "vdRctLCxSketdKb54xw67g",
    "version_id": 8000099,
    "version": "8.0.0",
    "indices": [],
    "data_streams": [],
    "include_global_state": false,
    "metadata": {
      "taken_by": "user123",
      "taken_because": "backup before upgrading"
    },
    "state": "SUCCESS",
    "start_time": "2020-06-25T14:00:28.850Z",
    "start_time_in_millis": 1593093628850,
    "end_time": "2020-06-25T14:00:28.850Z",
    "end_time_in_millis": 1593094752018,
    "duration_in_millis": 0,
    "failures": [],
    "shards": {
      "total": 0,
      "failed": 0,
      "successful": 0
    }
  }
}
----
// TESTRESPONSE[s/"uuid": "vdRctLCxSketdKb54xw67g"/"uuid": $body.snapshot.uuid/]
// TESTRESPONSE[s/"start_time": "2020-06-25T14:00:28.850Z"/"start_time": $body.snapshot.start_time/]
// TESTRESPONSE[s/"start_time_in_millis": 1593093628850/"start_time_in_millis": $body.snapshot.start_time_in_millis/]
// TESTRESPONSE[s/"end_time": "2020-06-25T14:00:28.850Z"/"end_time": $body.snapshot.end_time/]
// TESTRESPONSE[s/"end_time_in_millis": 1593094752018/"end_time_in_millis": $body.snapshot.end_time_in_millis/]
// TESTRESPONSE[s/"duration_in_millis": 0/"duration_in_millis": $body.snapshot.duration_in_millis/]
```

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:5]: By default, snapshots include an entire cluster, including the cluster state.
With #57675, snapshots now support data streams in `master` and `7.x`.

I'd avoid associating snapshots too closely with indices.

```suggestion
[[create-snapshot-api]]
=== Create snapshot API
++++
<titleabbrev>Create snapshot</titleabbrev>
++++
```

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:36]: I would add a higher-level intro sentence. We also need to mention data streams.
```suggestion
You can use the create snapshot API to create a <<snapshot-restore,snapshot>>, a
backup taken from a running {es} cluster.

By default, a snapshot includes all data streams and open indices in the
cluster, as well as the cluster state.  You can change this behavior by
specifying a list of data streams and indices to back up in the body of the
snapshot request.
```

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:40]: I made some suggestions to remove index references. These also apply to data streams now.
```suggestion
The snapshot process is incremental. When creating a  snapshot, {es} analyzes the list of files that are already stored in the repository and copies only files that were created or changed since the last snapshot. This process allows multiple snapshots to be preserved in the repository in a compact form.
```

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:42]:
```suggestion
The snapshot process is executed in non-blocking fashion, so all indexing and searching operations can run concurrently against the data stream or index that {es} is snapshotting. Only one snapshot process can run in the cluster at any time.
```

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:44]:
```suggestion
A snapshot represents a point-in-time view of the moment when the snapshot was created. No records that were added to a data stream or index after the snapshot process started will be present in the snapshot.
```

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:50]:
```suggestion
Besides creating a copy of each data stream and index, the snapshot process can also store global cluster metadata, including persistent cluster settings and templates. The transient settings and registered snapshot repositories are not stored as part of the snapshot.
```

**Review by jrodewig — CHANGES_REQUESTED** [review]
Thanks for putting this together @lockewritesdocs.

I left feedback that I think should be addressed before moving forward.

One important note: With #57675 merged, snapshots now support data streams. However, data streams are only available in the `master` and `7.x` branches.
If you decide to backport this to 7.8, you'll need to remove any data stream references from the backport.

I have a related PR (#58513) to update the existing snapshot/restore docs for data streams.

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (lockewritesdocs) [reply docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:26]: I was curious about whether to include the parameter, so thank you for the guidance.

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (lockewritesdocs) [reply docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:72]: I'll move `wait_for_completion` to the query params section.

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (lockewritesdocs) [reply docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:111]: Adding to the `metadata` description:
>Metadata must be less than 1024 bytes.

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (lockewritesdocs) [reply docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:107]: I'll add `partial` to the list of query parameters.

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (lockewritesdocs) [reply docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:112]: I'll add a description of `settings` with a pointer to that information in the [Put snapshot repository API docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/put-snapshot-repo-api.html).

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (lockewritesdocs) [reply docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc]: I'll make this change to request body params and also reorder them.

**Comment by lockewritesdocs** [comment]
@elasticmachine test this please

**Comment by lockewritesdocs** [comment]
@elasticmachine update branch

**Comment by lockewritesdocs** [comment]
@elasticmachine run docs-check

**Comment by lockewritesdocs** [comment]
@elasticmachine test this please

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:27]: It looks like the docs CI is failing because this snapshot doesn't finish before the example at the bottom. You can spoof the `wait_for_completion` query param using a `// TEST` comment.
```suggestion
[source,console]
-----------------------------------
PUT /_snapshot/my_repository/my_snapshot
-----------------------------------
// TEST[s/my_snapshot/my_snapshot?wait_for_completion=true/]
```

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:185]: I didn't notice this earlier, but it looks like this response includes a `version_id` and `version`. This won't cause problems on the `master` branch, but it'll fail as soon as it's ported to any other branches.

The suggestion below replaces the version properties with variables and fixes them with `// TESTRESPONSE` comments.
```suggestion
[source,console-result]
----
{
  "snapshot": {
    "snapshot": "snapshot_2",
    "uuid": "vdRctLCxSketdKb54xw67g",
    "version_id": <version_id>,
    "version": <version>,
    "indices": [],
    "data_streams": [],
    "include_global_state": false,
    "metadata": {
      "taken_by": "user123",
      "taken_because": "backup before upgrading"
    },
    "state": "SUCCESS",
    "start_time": "2020-06-25T14:00:28.850Z",
    "start_time_in_millis": 1593093628850,
    "end_time": "2020-06-25T14:00:28.850Z",
    "end_time_in_millis": 1593094752018,
    "duration_in_millis": 0,
    "failures": [],
    "shards": {
      "total": 0,
      "failed": 0,
      "successful": 0
    }
  }
}
----
// TESTRESPONSE[s/"uuid": "vdRctLCxSketdKb54xw67g"/"uuid": $body.snapshot.uuid/]
// TESTRESPONSE[s/"version_id": <version_id>/"version_id": $body.snapshot.version_id/]
// TESTRESPONSE[s/"version": <version>/"version": $body.snapshot.version/]
```

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:50]: Extra space. Line wrapping is also too long.
```suggestion
The snapshot process is incremental. When creating a snapshot, {es} analyzes the list of files that are already stored in the repository and copies only files that were created or changed since the last snapshot. This process allows multiple snapshots to be preserved in the repository in a compact form.
```

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (jrodewig) [inline docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:124]: I tested this parameter, and it seems weird... you can throw any arbitrary object in there and it accepts it.

@original-brownbear Any thoughts? I wonder if this is just a param that never got used. If so, we should maybe look at removing it.

In any case, I don't think this definition is accurate or makes sense. You shouldn't be able to overwrite snapshot repo settings from a snapshot.

If this is just an unused param, I wonder if we should A) not document it or B) document it with a warning that it doesn't do anything. I'd hate to see a user throw values in here expecting something to happen if it does nothing.

**Review by jrodewig — APPROVED** [review]
I have one important question about the `settings` parameter that I'll defer to @original-brownbear  on.  We should address that before merging.

Otherwise LGTM.

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (original-brownbear) [reply docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:124]: > If this is just an unused param, I wonder if we should A) not document it or B) document it with a warning that it doesn't do anything. I'd hate to see a user throw values in here expecting something to happen if it does nothing.

Sorry for the delay here. You're right ... this parameter doesn't do anything at the moment (I'll look into removing it). IMO, we probably should just not document it. Was it ever documented in the first place even?

**Review by original-brownbear — COMMENTED** [review]
As commented on inline, let's ditch the `settings` documentation

File: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc
Comment (original-brownbear) [reply docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc:124]: https://github.com/elastic/elasticsearch/pull/58661

Comment (lockewritesdocs) [reply docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc]: Thanks @jrodewig for testing this parameter, and @original-brownbear for confirming that we can remove it. With that parameter removed, I think that this PR is ready to merge.

Comment (jrodewig) [reply docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc]: Agree. Thanks for taking this one to the finish line @lockewritesdocs.

**Pull request merged**
